### PR TITLE
feat(hotkey): add a function test of hotkey detection

### PR DESCRIPTION
**Pull request by Smityz**
Add a function test of hotkey detection, and fix some bugs
1. Change the default value of `occurrence_threshold` `hot_bucket_variance_threshold` and `hot_key_variance_threshold` to improve stability.
2. Fix some logs and error hints.
3. Introduce `c_escape_string` to print not printable characters.

### CODE REVIEW

File: src/server/test/hotspot_partition_test.cpp [hunk 1]
```diff
@@ -126,6 +132,9 @@ TEST_F(hotspot_partition_test, hotspot_partition_policy)
 
 TEST_F(hotspot_partition_test, send_detect_hotkey_request)
 {
+    auto default_occurrence_threshold = FLAGS_occurrence_threshold;
+    FLAGS_occurrence_threshold = 100;
```
Comment (hycdong) [inline src/server/test/hotspot_partition_test.cpp:136]: Why do you update `occurrence_threshold` default value to 3 in `hotspot_partition_calculator.cpp`, and update it 100 in unit test?

File: src/test/function_test/test_detect_hotspot.cpp [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        system("./run.sh start_onebox -w");
+        chdir(global_env::instance()._working_dir.c_str());
+    }
+
+    void write_hotspot_data()
+    {
+        int64_t start = dsn_now_s();
+        int err = PERR_OK;
+        ASSERT_NE(pg_client, nullptr);
+
```
Comment (hycdong) [inline src/test/function_test/test_detect_hotspot.cpp:93]: // generate random data

File: src/test/function_test/test_detect_hotspot.cpp [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        const int64_t warmup_second = 30;
+        for (int i = 0; dsn_now_s() - start < warmup_second; ++i %= 1000) {
+            std::string index = std::to_string(i);
+            std::string h_key = generate_hash_key_by_random(false, 0);
+            std::string s_key = "sortkey_" + index;
+            std::string value = "value_" + index;
+            err = pg_client->set(h_key, s_key, value);
+            ASSERT_EQ(err, PERR_OK);
+        }
+
```
Comment (hycdong) [inline src/test/function_test/test_detect_hotspot.cpp:103]: // generate hotpot data

File: src/test/function_test/test_detect_hotspot.cpp [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        ASSERT_NE(pg_client, nullptr);
+
+        for (int i = 0; dsn_now_s() - start < max_detection_second; ++i %= 1000) {
+            std::string index = std::to_string(i);
+            std::string h_key = generate_hash_key_by_random(false, 0);
+            std::string s_key = "sortkey_" + index;
+            std::string value = "value_" + index;
+            err = pg_client->set(h_key, s_key, value);
+            ASSERT_EQ(err, PERR_OK);
+        }
```
Comment (hycdong) [inline src/test/function_test/test_detect_hotspot.cpp:169]: L162-L168 is similar with L94-L102, you can add another function for it.

File: src/test/function_test/test_detect_hotspot.cpp [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        bool find_hotkey = false;
+        int partition_index;
+        for (partition_index = 0; partition_index < partitions.size(); partition_index++) {
+            req.pid = dsn::gpid(app_id, partition_index);
+            auto errinfo =
+                ddl_client->detect_hotkey(partitions[partition_index].primary, req, resp);
+            ASSERT_EQ(errinfo, dsn::ERR_OK);
+            ASSERT_EQ(resp.err_hint,
+                      "Can't get hotkey now, now state: hotkey_collector_state::STOPPED");
+        }
```
Comment (hycdong) [inline src/test/function_test/test_detect_hotspot.cpp:188]: L171 -L188 is similar with L113, maybe you can add another function to reduce duplicated code.

File: src/test/function_test/test_detect_hotspot.cpp [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+};
+
+TEST_F(test_detect_hotspot, write_hotspot_data)
+{
+    std::cout << "start testing write_hotspot_data..." << std::endl;
+    write_hotspot_data();
+    write_random_data();
+    capture_until_maxtime();
+    read_hotspot_data();
+    read_random_data();
```
Comment (hycdong) [inline src/test/function_test/test_detect_hotspot.cpp:338]: The five functions above have many duplicated code, I recommend simply them.

File: src/server/test/hotspot_partition_test.cpp [hunk 7]
```diff
@@ -25,6 +25,7 @@ namespace pegasus {
 namespace server {
 
 DSN_DECLARE_int32(occurrence_threshold);
+DSN_DECLARE_int32(enable_detect_hotkey);
```
Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:28]: `DSN_DECLARE_bool`

Comment (levy5307) [inline src/server/test/hotspot_partition_test.cpp:28]: DSN_DECLARE_bool

File: src/server/hotkey_collector.cpp [hunk 8]
```diff
@@ -239,23 +241,27 @@ void hotkey_collector::on_start_detect(dsn::replication::detect_hotkey_response
...
-        resp.err = dsn::ERR_INVALID_STATE;
-        hint = fmt::format(
-            "{} hotkey result has been found, you can send a stop rpc to restart hotkey detection",
-            dsn::enum_to_string(_hotkey_type));
+        resp.err = dsn::ERR_BUSY;
+        hint = fmt::format("{} hotkey result has been found: {}, you can send a stop rpc to "
+                           "restart hotkey detection",
+                           dsn::enum_to_string(_hotkey_type),
+                           pegasus::utils::c_escape_string(_result.hot_hash_key));
+        resp.__set_err_hint(hint);
```
Comment (levy5307) [inline src/server/hotkey_collector.cpp:257]: you can move these two lines out of switch:
```
        resp.__set_err_hint(hint);
        dwarn_replica(hint);
```

File: src/server/hotkey_collector.cpp [hunk 9]
```diff
@@ -265,6 +265,8 @@ void hotkey_collector::on_start_detect(dsn::replication::detect_hotkey_response
         derror_replica(hint);
         dassert(false, "invalid collector state");
     }
+    resp.__set_err_hint(hint);
+    dwarn_replica(hint);
```
Comment (hycdong) [inline src/server/hotkey_collector.cpp:269]: L268-L269 won't be executed, because it will either return or assert before it.

**Review by hycdong — APPROVED** [review]

**Review by levy5307 — APPROVED** [review]

**Pull request merged**